pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go

```diff
@@ -101,17 +101,19 @@ var _ = Describe("Qemu agent poller", func() {
 
 			go w.Poll(func(commands []AgentCommand) { done <- struct{}{} }, c)
```

Comment:
i am afraid from this logic
```
			c := make(chan struct{})
			defer close(c)
			done := make(chan struct{})
			defer close(done)
```
Once `runPollAndCountCommandExecution` finishes 
it will close the 2 channels, 
if `w.Poll` will enter its own timer and try to send to `done` channel, it will panic
sending on closed channel

Reply:
`countSignals` is waiting on the `done` channel, counting signals until the timeout.
When the timeout is reached, it exits with the counting, then closes the channels. We add a 10msec buffer for the timeout in order to timeout after the last expected tick. This gives ~990msec until the next attempt to send the signal (assuming the minimal 1sec tick), which is more than enough not to encounter the problem of sending a signal on a closed channel.

I would avoid referencing informal documentations/tutorials when a formal [one](https://tour.golang.org/flowcontrol/13) is available.
And it does not matter who is closed first, you cannot be certain the go-routine will "see" the signal before we close the other one as well, then it is a matter who is processed first (the tick signal or the close one).

But as this is bothering you so much and I find it fun to play with these channels in unit tests, I can just remove the redundant `done` channel close instruction. Closing channel `c` assures the go-routine is cleaned up and therefore `done` is left without reference and will be collected by the GC eventually.

Reply:
What about looking from the "left" side of the time line, not on the 990msec left, but having timers jitter where the timers were faster than they should and then we are left marginal.
I prefer code which is robust to jitters, for example the suggestions we talked previously, but its up to you.
The minimal step i would suggest is reordering the closing channels, its a good practice of the code, to make it as right as possible, and then when the right time comes we will more likely write it right, again up to you.

Reply:
> the timers were faster than they should

First time I hear about timers returning faster and not latter.

> for example the suggestions we talked previously

There have been no good suggestions or practical ones so far. If you have one, you are welcome to suggest it.

> i would suggest is reordering the closing channels, its a good practice of the code, to make it as right as possible, and then when the right time comes we will more likely write it right

It is good practice to write something correct, the suggestion of reordering is not correct as explained in my previous comment. It will not make anything better than now for sure.

> What you said is also good of course.

If that is correct, why do you insist on the other incomplete approaches? I do not understand.

> There are solutions where it wouldn't happen, as we discussed, and they are preferred.

Sorry, but no practical solutions have been given by you or other reviewers. I provided a 2nd approach that uses signals over sleep. If you have a better solution, suggest it, hopefully with a code snippet.

> again up to you.

You are blocking your review on these points, so repeating this is not assisting me to convince the reviewers that the tests are good enough.

---

I am going to remove the `done` channel close so the GC will pick it up and explicitly make sure no one will send on a closed channel.
If you have additional comments on how to improve it, please be specific and avoid generalizations so we can push this change forward.

Reply:
1. I am not blocking, i didn't even remove the lgtm, just raised my concerns.
2. I gave a bullet proof solution, waiting that the timer stop raising as indication that the go stopped after eventually it reached the desired value, and then you aint play with jitters at all.

Reply:
>     2\. I gave a bullet proof solution, waiting that the timer stop raising as indication that the go stopped after eventually it reached the desired value, and then you aint play with jitters at all.

So you use a buffer as well, just a huge one which is dependent on the tick (making the next test take tens of seconds). And yes, you play with the jitter, you just use an huge buffer and will probably decide on a timeout which is much bigger than the one suggested in the current solution.

Reply:
I find it personally better, and it shouldn't take 10 seconds, and even there are buffers / jitter, they are embedded, and not explicitly which is better imo, because of safely and readability.

Eventually reach N (with safe timeout - usually will take much less than a second extra)
close channel
Eventually Doesn't change (with safe timeout - usually will take up to 1 second)

So the buffer is 1 second + delta more than what you need.

Reply:
> I find it personally better, and it shouldn't take 10 seconds, and even there are buffers / jitter, they are embedded, and not explicitly which is better imo, because of safely and readability.
> 
> a. Eventually reach N (with safe timeout (as Eventually argument) - usually will take much less than a second extra)
> b. Close channel
> c. Eventually Doesn't change (with safe timeout (as Eventually argument) - usually will take up to 1 second)
> 
> So the buffer is 1 second + delta more than what you need at most.
> and it checks end 2 end including the Poll go routine wrap up.

I disagree, the polling is a mess and requires the same amount of timeout+bugger logic. If you think a polling approach is better than an event approach, we have no agreement.